Multiple Spectrometers

KatieKoch2 opened the pull request:
To test this code, make sure that multiple spectrometers are connected to the computer. 

To switch between spectrometers make sure the program is not actively taking spectra, then select the spectrometer you want from the drop-down menu in the quick view area of the GUI. Test that the spectra are acquired by using the view button. Again, remember to stop the acquisition before switching to a different spectrometer.
<img width="1380" height="907" alt="Screenshot 2025-11-18 at 9 38 33 AM" src="https://github.com/user-attachments/assets/f9dffcdf-be63-4dd2-8b90-58530cfe7ae4" />
<img width="1380" height="907" alt="Screenshot 2025-11-18 at 9 38 12 AM" src="https://github.com/user-attachments/assets/ae718d49-12de-4b36-b486-cea592c211f0" />
<img width="1380" height="907" alt="Screenshot 2025-11-18 at 9 37 59 AM" src="https://github.com/user-attachments/assets/b049c3cc-2dd7-436d-91b3-7114b6297a12" />

KatieKoch2 committed:
Update main.py

KatieKoch2 committed:
Update Instruments.py

KatieKoch2 committed:
Update Stresing.py

KatieKoch2 committed:
Update Stresing.py

KatieKoch2 committed:
Update StresingDriver.py

KatieKoch2 committed:
Update DataHandling.py

KatieKoch2 committed:
Update main_GUI.ui

fthouin commented:
Thanks a lot Katie! We will test this one ourselves. @SDaneau we could use a OceanOptics to directly test it, otherwise we could check if we can configure the code to work with SilvaBot by copying a few lines of code from that repo?

KatieKoch2 committed:
Update contributing.md

fthouin commented on manual/contributing.md:
```diff
@@ -65,3 +65,7 @@ To calibrate the SLM for various wavelengths, one needs to create a Phase to Gra
 This is done in two steps:
 1. Measure_LUT_PhasetoGreyscale: Displays a pattern on the SLM where half of the SLM is set to a grayscale value of zero, and the other half of the SLM scans through the grayscale values (0-255). The spectrum of the beam is taken after each pattern. 
 2. Generate_LUT_PhasetoGreyscale: Analyzes the measured spectrum to determine the phase shift from the reference (where both sides of the SLM are at a greyscale value of zero). This is done by taking a Fourier transform of the spectrum and calculating the phase difference using the real and imaginary components. 
+
+### Spectrometer Selection 
+
+This feature allows multiple spectrometers to be stored within a dictionary class, 'spectrometers', so that the user can switch between the active spectrometer. There is a drop down menu on the quick control section of the GUI. When a different spectrometer is selected from the dropdown menu, it updates the device dictionary with the active spectrometer and changes the associated spectral length  ...
```
This section would be more relevant in another page of the manual. You could include it in index.md under the Hardware section or making a new page

fthouin commented on src/DataHandling/DataHandling.py:
```diff
@@ -117,17 +118,24 @@ def concatenate_data(self, wls, spec):
+                spec = spec - self.background.ravel()
+                # print('bg shape', np.shape(self.background))
+                # print('spec shape', np.shape(spec))
+
```
Can you remove the commented prints?

fthouin commented on src/DataHandling/DataHandling.py:
```diff
@@ -252,6 +260,38 @@ def load_data(self):
+        self.speclength = new_length
+
+        # Reset spectrum buffer
+        self.spec = np.zeros((self.speclength, 0))  # zero columns, rows = new_length
```
Please remove commented code lines that are not used anymore.

fthouin commented on src/DataHandling/DataHandling.py:
```diff
@@ -252,6 +260,38 @@ def load_data(self):
+        # preallocate data arrays depending on data dimension (1D or 2D).
+        if self.data_dim == 1:
+            self.spec = np.empty([self.speclength, 0])
+            #print('1D', self.spec)
```
Could you remove this commented print?

fthouin commented on src/DataHandling/DataHandling.py:
```diff
@@ -252,6 +260,38 @@ def load_data(self):
+            self.wls = np.empty([self.speclength, 1])
+        else:
+            self.spec = np.empty([0, self.speclength[0], self.speclength[1]])
+            #print('2D', self.spec)
```
Could you remove this commented print?

fthouin commented on src/main.py:
```diff
@@ -132,6 +133,32 @@ def __init__(self):
         
+        #Spectrometer Selection
+        self.spectrometer_select = self.findChild(QtWidgets.QComboBox, 'spec_selection_comboBox')
+        # print('Spec Select', self.spectrometer_select)
```
Could you remove this commented print?

fthouin commented on src/main.py:
```diff
@@ -132,6 +133,32 @@ def __init__(self):
+
+        if self.spectrometer_select is not None:
+            logger.info(f"Available spectrometers: {list(self.spectrometers.keys())}")
+            # print("Available spectrometers:", self.spectrometers)
```
Could you remove this commented print?

fthouin commented on src/main.py:
```diff
@@ -132,6 +133,32 @@ def __init__(self):
+
+        self.active_spectrometer = self.devices['spectrometer']
+        logger.info(f"Available Devices: {list(self.devices.keys())}")
+        # print('Available devices:', self.devices)
```
Could you remove this commented print?

fthouin reviewed (changes requested):
Thanks for the contribution Katie! A couple of formatting comments below. I'll let Simon confirm that the code behaves as intended in the lab.

SDaneau commented:
The program seems to run correctly. However, I'm expecting some problem with the actual dev branch, i.e. the save and load buttons are located in the quick control panel and may be over the drop-down menu. I suggest that you merge dev in that branch first and make sure everything is fine.

KatieKoch2 committed:
Update DataHandling.py

KatieKoch2 committed:
Merge remote-tracking branch 'origin/dev' into Spec_Selection_KK

KatieKoch2 committed:
Merge branch 'Spec_Selection_KK' of https://github.com/SilvaScience/colberto into Spec_Selection_KK

KatieKoch2 committed:
Update main.py

KatieKoch2 committed:
Update main_GUI.ui

KatieKoch2 committed:
Update Instruments.py

KatieKoch2 committed:
Update StresingDriver.py

KatieKoch2 committed:
Update Stresing.py

KatieKoch2 committed:
Update DataHandling.py

KatieKoch2 committed:
Update DataHandling.py

KatieKoch2 committed:
Update Instruments.py

KatieKoch2 committed:
Update main.py

KatieKoch2 committed:
Update Stresing.py

KatieKoch2 commented:
I merged dev into the branch and updated the code so that it runs properly. While doing this I realized I missed a key component when integrating this feature, which is updating the parameter tree since the camera has more parameters than the Ocean or DEMO spec. Now, when the spectrometer is changed, the spectrometer tab in the side bar is also updated.

KatieKoch2 committed:
Update DataHandling.py

KatieKoch2 committed:
Update DataHandling.py

KatieKoch2 committed:
Update main.py

SDaneau commented:
@fthouin, I tried that branch in the lab and everything seems to be fine! I think we can merge. Bravo @KatieKoch2
